litr/src/main/java/com/linkedin/android/litr/render/AudioRenderer.kt

```diff
@@ -0,0 +1,134 @@
+
+    override fun init(outputSurface: Surface?, sourceMediaFormat: MediaFormat?, targetMediaFormat: MediaFormat?) {
+        onMediaFormatChanged(sourceMediaFormat, targetMediaFormat)
+        released.set(false)
+        renderThread.start()
+    }
+
+    override fun onMediaFormatChanged(sourceMediaFormat: MediaFormat?, targetMediaFormat: MediaFormat?) {
+        this.sourceMediaFormat = sourceMediaFormat
+        this.targetMediaFormat = targetMediaFormat
+
+        if (targetMediaFormat?.containsKey(MediaFormat.KEY_SAMPLE_RATE) == true) {
+            sampleDurationUs = 1_000_000f / targetMediaFormat.getInteger(MediaFormat.KEY_SAMPLE_RATE)
+        }
+
+        if (targetMediaFormat?.containsKey(MediaFormat.KEY_CHANNEL_COUNT) == true) {
+            channelCount = targetMediaFormat.getInteger(MediaFormat.KEY_CHANNEL_COUNT)
+        }
+    }
+
+    override fun getInputSurface(): Surface? {
+        return null
+    }
+
+    override fun renderFrame(inputFrame: Frame?, presentationTimeNs: Long) {
+        if (!released.get() && inputFrame?.buffer != null) {
+            val buffer = ByteBuffer.allocate(inputFrame.buffer.limit())
+            buffer.put(inputFrame.buffer)
+            buffer.flip()
+
+            val bufferInfo = MediaCodec.BufferInfo()
+            bufferInfo.set(
+                0,
+                inputFrame.bufferInfo.size,
+                inputFrame.bufferInfo.presentationTimeUs,
+                inputFrame.bufferInfo.flags
+            )
+
+            renderQueue.add(Frame(inputFrame.tag, buffer, bufferInfo))
+        }
+    }
+
+    override fun release() {
+        released.set(true)
+    }
+
+    override fun hasFilters(): Boolean {
+        return false
+    }
+
+    private inner class RenderThread : Thread() {
+        override fun run() {
+            while (!released.get()) {
+                renderQueue.peekFirst()?.let { inputFrame ->
+                    val tag = encoder.dequeueInputFrame(FRAME_WAIT_TIMEOUT)
+                    when {
+                        tag >= 0 -> renderFrame(tag, inputFrame)
+                        tag == MediaCodec.INFO_TRY_AGAIN_LATER -> {} // do nothing, will try later
+                        else -> Log.e(TAG, "Unhandled value $tag when receiving decoded input frame")
+                    }
+                }
+            }
+            renderQueue.clear()
+        }
+
+        private fun renderFrame(tag: Int, inputFrame: Frame) {
+            encoder.getInputFrame(tag)?.let { outputFrame ->
+                if (outputFrame.buffer != null && inputFrame.buffer != null) {
+                    outputFrame.bufferInfo.offset = 0
+                    outputFrame.bufferInfo.flags = inputFrame.bufferInfo.flags
+                    outputFrame.bufferInfo.presentationTimeUs =
+                        inputFrame.bufferInfo.presentationTimeUs +
+                            ((inputFrame.buffer.position() / (channelCount * BYTES_PER_SAMPLE)) * sampleDurationUs).toLong()
```

Comment:
BYTES_PER_SAMPLE could be 8, 16, 32, right?

Reply:
No, on Android it is only 16. MediaCodec uses Short or recently 16 bit Float. There is support for 24 bit audio, but only though some audio specific APIs, not MediaCodec.